presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -630,6 +675,34 @@ protected boolean internalStartNext()
         }
     }
 
+    protected void enforceTimeLimits()
+    {
+        checkState(Thread.holdsLock(root), "Must hold lock to enforce time limits");
+        synchronized (root) {
+            for (InternalResourceGroup group : subGroups.values()) {
+                group.enforceTimeLimits();
+            }
+            long runningTimeoutMillis = runningTimeout.toMillis();
+            long queuedTimeoutMillis = queuedTimeout.toMillis();
+            // Copy of activeQueries since queryFinished gets called by cancelQuery which modifies activeQueries
+            for (QueryExecution query : ImmutableList.copyOf(activeQueries)) {
+                if (runningQueries.contains(query)) {
+                    if (Optional.ofNullable(query.getQueryInfo().getQueryStats().getExecutionTime())
+                            .orElse(new Duration(0, NANOSECONDS)).toMillis() > runningTimeoutMillis) {
+                        cachedMemoryUsageBytes -= query.getTotalMemoryReservation();
+                        query.fail(new PrestoException(EXCEEDED_TIME_LIMIT, "query exceeded resource group runtime limit"));
+                    }
+                } // Queued time may be null when a query is queued, subtract execution time from elapsed time instead
```

Comment:
Move to next line.
I think queued time is *always* null when the query is queued.
Just pull out the 'queued' time into a variable - make your conditionals simpler.